src/vision/status_quo/barbara_compares_some_cpp_code.md

```diff
@@ -0,0 +1,111 @@
+
+## The story
+
+Barbara is recreating some code that has been written in other languages they have some familiarity with. These include C++, but
+also GC'd languages like Python.
+
+This code collates a large number of requests to network services, with eash response containing a large amount of data.
+To speed this up, Barbara uses `buffer_unordered`, and writes code like this:
+
+```
+let mut queries = futures::stream::iter(...)
+    .map(|query| async move {
+        let d: Data = self.client.request(&query).await?;
+        d
+     })
+     .buffer_unordered(32);
+
+use futures::stream::StreamExt;
+let results = queries.collect::<Vec<Data>>().await;
+```
+
+Barbara thinks this is similar in function to things she has seen using
+Python's [asyncio.wait](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait),
+as well as some code her coworkers have written using c++20's `coroutines`,
+using [this](https://github.com/facebook/folly/blob/master/folly/experimental/coro/Collect.h#L321):
+
+```
+std::vector<folly::coro::Task<Data>> tasks;
+ for (const auto& query : queries) {
+    tasks.push_back(
+        folly::coro::co_invoke([this, &query]() -> folly::coro::Task<Data> {
+              co_return co_await client_->co_request(query);
+        }
+    )
+}
+auto results = co_await folly:coro::collectAllWindowed(
+      move(tasks), 32);
+```
+
+However, *the Rust code performs quite poorly compared to the other impls,
+appearing to effectively complete the requests serially, despite on the surface
+looking like effectively identical code.*
+
+Barbara goes deep into investigating this, spends time reading how `buffer_unordered` is
+implemented, how its underlying `FuturesUnordered` is implemented, and even thinks about
+how polling and the `tokio` runtime she is using works. She evens tries to figure out if the
+upstream service is doing some sort of queueing.
+
+Eventually Barbara starts reading more about c++20 coroutines, looking closer at the folly
+implementation used above, noticing that is works primarily with *tasks*, which are not exactly
+equivalent to rust `Future`'s.
+
+Then it strikes her! `request` is implemented something like this:
+```
+impl Client {
+    async fn request(&self) -> Result<Data> {
+        let bytes = self.inner.network_request().await?
+        Ok(serialization_libary::from_bytes(&bytes)?)
+   }
+}
+```
+The results from the network service are VERY large, and the `BufferedUnordered` stream is contained within 1 tokio task.
+**The request future does non-trivial cpu work to deserialize the data.
+This causes significant slowdowns in wall-time as the the process CAN BE bounded by the time it takes
+the single thread running the tokio-task to deserialize all the data.**
```

Comment:
Change the story so that it's clearer why Barbara didn't notice her mistake immediately?
```suggestion
The results from the network service are sometimes (but not always) VERY large, and the `BufferedUnordered` stream is contained within 1 tokio task.
**The request future does non-trivial cpu work to deserialize the data.
This causes significant slowdowns in wall-time as the the process CAN BE bounded by the time it takes
the single thread running the tokio-task to deserialize all the data.**

This problem hadn't shown up in test cases, where the results from the mocked network service are always small; many common uses of the network service only ever have small results, so it takes a specific production load to trigger this issue.
```